File: python/calendar.py
import argparse
from datetime import datetime, timedelta
import sys


class Calendar:
    WEEK_DAYS = ["Mo", "Tu", "We", "Th", "Fr", "Sa", "Su"]
    HIGHLIGHT_CODE = "\033[7m"
    RESET_CODE = "\033[0m"

    def __init__(self, year, month):
        self.year = year
        self.month = month

    def get_first_day(self):
        return datetime(self.year, self.month, 1)

    def get_last_day(self):
        first_day = self.get_first_day()
        return (first_day.replace(day=28) + timedelta(days=4)).replace(
            day=1
        ) - timedelta(days=1)

    def get_year_month_string(self):
        return self.get_first_day().strftime("%B %Y")

    def format_day(self, current_day, today):
        if current_day.date() == today:
            return f"{self.HIGHLIGHT_CODE}{current_day.day:2d}{self.RESET_CODE}"
        else:
            return f"{current_day.day:2d}"

    def generate_first_week(self, current_day, today):
        SUNDAY = 6
        # 月の最初の週の空白部分を埋める
        first_week = ["  "] * current_day.weekday()

        while current_day.weekday() != SUNDAY:
            day_str = self.format_day(current_day, today)
            first_week.append(day_str)
            current_day += timedelta(days=1)

        day_str = self.format_day(current_day, today)
        first_week.append(day_str)
        current_day += timedelta(days=1)
        return first_week, current_day

    def generate_week(self, current_day, last_day, today):
        week = []

        while len(week) < 7 and current_day <= last_day:
            day_str = self.format_day(current_day, today)
            week.append(day_str)
            current_day += timedelta(days=1)
        return week, current_day

    def generate_calendar(self):
        first_day = self.get_first_day()
        last_day = self.get_last_day()
        year_month = self.get_year_month_string()
        width = len(" ".join(self.WEEK_DAYS))

        calendar_lines = [
            year_month.center(width),
            " ".join(self.WEEK_DAYS),
        ]

        current_day = first_day
        today = datetime.now().date()

        first_week, current_day = self.generate_first_week(current_day, today)
        calendar_lines.append(" ".join(first_week))

        while current_day <= last_day:
            week, current_day = self.generate_week(current_day, last_day, today)
            calendar_lines.append(" ".join(week))

        return "\n".join(calendar_lines)


class CustomArgParser(argparse.ArgumentParser):
    def error(self, message):
        message = message.replace("argument -m/--month: ", "")
        print(message, file=sys.stderr)
        sys.exit(2)


def validate_month(month):
    if not month.isdigit() or not (1 <= int(month) <= 12):
        raise argparse.ArgumentTypeError(
            f"{month} is neither a month number (1..12) nor a name"
        )
    return int(month)


def parse_arguments():
    parser = CustomArgParser(description="Display a calendar for a specific month")
    parser.add_argument(
        "-m",
        "--month",
        type=validate_month,
        help="Month to display (1-12)",
        default=datetime.now().month,
    )

    return parser.parse_args()


def main():
    try:
        args = parse_arguments()
        now = datetime.now()
        c = Calendar(now.year, args.month)
        calendar = c.generate_calendar()
        print(calendar)
    except Exception as e:
        print(f"An unexpected error occurred: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: js/calendar.js
class Calendar {
        static DAYS_IN_WEEK = 7;

        static DISPLAY = {
                SPACE_PER_DAY: 3, // 日付の表示幅（2桁 + 空白1文字）
                DATE_PADDING: 2, // 日付を2桁で右寄せ表示するためのパディング幅
        };

        static MONTH_NAMES = [
                "1月",
                "2月",
                "3月",
                "4月",
                "5月",
                "6月",
                "7月",
                "8月",
                "9月",
                "10月",
                "11月",
                "12月",
        ];

        // 指定された年月の1日の曜日を取得
        static getFirstDay(year, month) {
                return new Date(year, month - 1, 1).getDay();
        }

        // 指定された年月の最終日を取得
        static getLastDate(year, month) {
                return new Date(year, month, 0).getDate();
        }

        static generate(year, month) {
                const firstDay = Calendar.getFirstDay(year, month);
                const lastDate = Calendar.getLastDate(year, month);

                // ヘッダー部分の生成（年月と曜日）
                let calendar = `      ${
                        Calendar.MONTH_NAMES[month - 1]
                } ${year}\n`;
                calendar += "日 月 火 水 木 金 土\n";

                // カレンダー本体の生成
                let currentDay = 1;
                let line = "".padStart(
                        firstDay * Calendar.DISPLAY.SPACE_PER_DAY,
                );

                for (let i = firstDay; currentDay <= lastDate; i++) {
                        // i > 0：最初の週での不要な改行を防ぐ（firstDayが0の場合）
                        if (i > 0 && i % Calendar.DAYS_IN_WEEK === 0) {
                                calendar += line.trimEnd() + "\n";
                                line = "";
                        }
                        line += currentDay.toString().padStart(
                                Calendar.DISPLAY.DATE_PADDING,
                        ) + " ";
                        currentDay++;
                }

                // 最終行の処理
                if (line) {
                        calendar += line.trimEnd() + "\n";
                }
                return calendar;
        }
}

// -m オプションで月の指定が可能。省略時はmain()でシステムの現在の月が使用される
function parseArgs() {
        const args = process.argv.slice(2);
        const options = { month: null };

        if (args.length === 0) return options;

        if (args[0] !== "-m" || args.length !== 2) {
                throw new Error("Usage: node calendar.js [-m month]");
        }

        const month = parseInt(args[1]);
        if (isNaN(month) || month < 1 || month > 12) {
                throw new Error("月は1から12の整数で指定してください");
        }

        options.month = month;
        return options;
}

function main() {
        try {
                const options = parseArgs();
                const currentDate = new Date();
                const month = options.month || currentDate.getMonth() + 1;
                const year = currentDate.getFullYear();
                process.stdout.write(Calendar.generate(year, month));
        } catch (error) {
                console.error(`Error: ${error.message}`);
                process.exit(1);
        }
}

main();
